### ISSUE TITLE
Dependency Conflict on unimplemented method is not showing up as a cause

### ISSUE THREAD

**Issue by suztomo** [GoogleCloudPlatform/cloud-opensource-java#1537]
https://gist.github.com/suztomo/e40eb79989d20317eb61784a9c8de654

```
(org.apache.hadoop:hadoop-core:0.20.2) org.apache.hadoop.hdfs.server.namenode.FSEditLog$EditLogFileInputStream's method isInProgress() is not implemented in the class;
  referenced by 1 class file
    org.apache.hadoop.hdfs.server.namenode.EditLogInputStream (org.apache.hadoop:hadoop-hdfs:2.6.0)
  Cause:
    Unknown
```

**Comment by elharo** [comment]
I don't believe this is accurate. E.g. "The auto-value users' projects will not have incap-processor artifacts at runtime." Autovalue is a weird one that shouldn't be in project classpaths at all. However in general a provided dependency **is** in a project's compile time and runtime classpaths. Maven doesn't put it there, but it comes from somewhere else.

**Comment by suztomo** [comment]
I understand that there're cases that provided dependencies come from somewhere.

But there are other cases where provided dependencies are not needed to be present at runtime. For example users of `org.apache.beam:beam-sdks-java-io-hcatalog:jar:2.19.0` do not need `org.apache.hadoop:hadoop-core:0.20.2` at runtime.

**Comment by suztomo** [comment]
David: not to report if there's a provided depency between the root and the the linkage error's source artifact.

We haven't decided action yet.

**Comment by elharo** [comment]
I wonder if that's an abuse of `provided`. Per Maven docs provided "is much like compile, but indicates you expect the JDK or a container to provide the dependency at runtime. For example, when building a web application for the Java Enterprise Edition, you would set the dependency on the Servlet API and related Java EE APIs to scope provided because the web container provides those classes. A dependency with this scope is added to the classpath used for compilation and test, but not the runtime classpath. It is not transitive."

I'll ask the Maven team what they think about using provided in place of compileOnly. Is this blessed?

**Comment by suztomo** [comment]
Nice. I'm interested in what Maven folks think (not sure how match context you can deliver this static analysis tool).

Yes, abuse of provided is good term; such libraries can work fine without the presence of the provided dependencies.
